### DISCUSSION TITLE
Changing database unit

### DISCUSSION THREAD

**Discussion by fp-matt** [gdsfactory/gdsfactory#2783] [Q&A] [answered]
I don't really know a whole lot about KFactory, but it seems like the default layout database unit is different than it was in gdstk (i.e., nm instead of um). 

This has been causing some really weird behavior in my layout. Such as how the components draw as they did before (in units of um), but pins now draw as if the shape was defined on the nm scale and then scaled to match the um scale, which means that my pins are 1000x larger than my parts now. Additionally, when I've been trying to translate component references, they only move on the nm scale.

For example, trying to a component 5um to the right and 50um down,
![image](https://github.com/gdsfactory/gdsfactory/assets/169110167/bb47e4fe-c0e7-46fa-903d-77496af1248a)

or trying to add a pin to a component that is 100 um wide
![image](https://github.com/gdsfactory/gdsfactory/assets/169110167/6f425bb1-a80a-4064-a92b-241777a95b80)

Is there an easy way to globally change the dbu? It's hard to know all of the ways that things have changed between gdstk and KFactory, so I'm not really sure if there's something that I am specifically doing wrong.

**Comment by tvt173** [comment]
It's not that the dbu size has changed, it's that klayout operates on integer dbu by default and the gdsfactory interface has always been consistently microns. I agree that this is a confusing and jarring change, and we've been talking about finding a way around it before we release 8.0. Right now, you must call `my_ref.d.move(...)` rather than `my_ref.move(...)` for example to do the operation in microns. I think it would be great if we can make it the opposite... Do all operations consistently in microns by default, as before, unless specifically requesting to do them in integer dbu

**Reply by fp-matt** [reply]
I think the most confusing part of the change for me is that it isn't consistent whether an operation/component creation is done in microns or dbu.

Also, while this makes sense with regards to moving components around, I'm still not sure this answers the issue I ran into with respect to the pins being drawn at 1000x scale. 

Also I completely forgot to mention that to actually even get the `add_pins_triangle(...)` function to work, I had to edit the `Component.add_label(...)` method to accept `position` as an instance of np.ndarray because of the return type of the `get_pin_triangle_polygon_tip(...)` function. Typing this out, I'm realizing that I may have inadvertently created the scaling issue by hacking together the `Component.add_label(...)` method, but wanted to make mention of it here still just in case.

**Comment by joamatab** [comment]
yes, you can take a look at the migration guideline we just published here

https://gdsfactory.github.io/gdsfactory/notebooks/21_migration_guide_7_8.html

**Reply by fp-matt** [reply]
Ah okay, I hadn't seen the migration guide yet - After I finished working my way through the layout tutorial a while back, I've mostly been referencing the docstrings in the source code instead of the github page

**Comment by joamatab** [comment]
can we subclass to keep the old behavior?

```python

def __getattribute__(self, __k: str) -> Any:
     if __k in ["xmin","xmax","xsize","ysize","x","y","center","create_inst":
         return getattr(self.d,__k)

```

@sebastian-goeldi

**Reply by sebastian-goeldi** [reply, marked as answer]
Yes this is possible. I'll finish up the [PR](https://github.com/gdsfactory/gdsfactory/pull/2785) as soon as I can get to it.

Also regarding the DBU changing, yes it's possible to do so as long as it happens at the very beginning of any script:

`gf.kcl.dbu = new_dbu_value`  where `new_dbu_value` is the relation of `um` -> `dbu`. The standard is that 1dbu == 1nm, so the `gf.kcl.dbu==0.001`.

**Reply by nwasserb** [reply]
@sebastian-goeldi 

I can't seem to get the gf.kcl.dbu = new_dbu_value part to work when changing DBU units. Can you provide a bit more guidance on steps towards changing the dbu grid size?

**Reply by sebastian-goeldi** [reply]
@nwasserb I am not sure I fully follow.

The following works for me without problems:

```python
import gdsfactory as gf


gf.kcl.dbu = 0.005 # change dbu to 5nm, it's always in relation to 1um

print(gf.kf.kcl.layout.dbu) # this is where klayout internally has the dbu, so this is to check that it's actually there where it should be
```

**Reply by l1gh7b3nd3r** [reply]
@sebastian-goeldi Hi, I was trying to do the same. I would like to reduce my DBU less than 1 nm. I tried your code, but it is returning me this error.

Traceback (most recent call last):
  File "/home/anoop/Desktop/code/mask/trial.py", line 4, in <module>
    gf.kcl.dbu = 0.005 # change dbu to 5nm, it's always in relation to 1um
AttributeError: module 'gdsfactory' has no attribute 'kcl'
Could you please tell me what might be the possible solution?

**Reply by sebastian-goeldi** [reply]
Are you sure you are using gdsfactory8?
7 will not have a `gf.kcl`

**Reply by l1gh7b3nd3r** [reply]
I was not using v8. Thank You.
